File: AboutBoxWx/mainframe.cpp
#include "mainframe.h"

#include "main.h"
#include "aboutlogo.h"
#include "appversion.h"

#include <wx/aboutdlg.h>

#ifndef __WXMSW__
#include "wxwin32x32.xpm"
#endif // __WXMSW__

const wxChar* g_szLicence = _T(
"This application is provided \"as is\" without any licence\n"
"\n"
"Everyone is free to copy, modify, sell, and do anything with it.\n"
"\n"
"This application and its source code is provided as a small\n"
"sample to show the use of wxWidgets standard about box."
);

const wxChar* g_szTranslators[] = {
	_T("Xaviou (en = app default)"),
	_T("Nobody Else (no translations support)")
};

const int g_iCheckedBoxes[] = {
	CHK_APP_NAME,
	CHK_APP_VERSION,
	CHK_APP_DESCRIPTION,
	CHK_COPYRIGHT,
	CHK_ICON
};

MainFrame::MainFrame(const wxString& title) : wxFrame(NULL, -1, title)
{
#ifdef __WXDEBUG__
	wxPrintf(_T("Creating a \"MainFrame\" object\n"));
#endif // __WXDEBUG__

	SetIcon(wxICON(appIcon)); // Defining app icon

	CreateControls();

	ConnectControls();

	UpdateInterface();
}

MainFrame::~MainFrame()
{
#ifdef __WXDEBUG__
	wxPrintf(_T("Destroying a \"MainFrame\" object\n"));
#endif // __WXDEBUG__
}

void MainFrame::CreateControls()
{
	// Status bar
	CreateStatusBar(1);
	wxString sTxt=wxGetApp().GetBuildInfos(true);
	SetStatusText(sTxt);

	// Add a panel to the frame (needed under Windows to have a nice frame)
	wxPanel *pnl=new wxPanel(this, -1);
	wxBoxSizer *szrMain=new wxBoxSizer(wxVERTICAL);
		wxStaticBoxSizer *stbSzr=new wxStaticBoxSizer(wxVERTICAL, pnl, _T("Informations to show in the \"About\" box:"));

			m_optInfos[OPT_MINIMAL] = new wxRadioButton(pnl, wxID_ANY, _T("Application's name only (minimal \"Aboud\" box)"), wxDefaultPosition, wxDefaultSize, wxRB_GROUP);
			stbSzr->Add(m_optInfos[0], 0, wxALL, 5);

			m_optInfos[OPT_MAXIMAL] = new wxRadioButton(pnl, wxID_ANY, _T("All available informations"));
			stbSzr->Add(m_optInfos[1], 0, wxLEFT|wxRIGHT|wxBOTTOM, 5);

			m_optInfos[OPT_CUSTOM] = new wxRadioButton(pnl, wxID_ANY, _T("Let me select the informations I want to see"));
			stbSzr->Add(m_optInfos[2], 0, wxLEFT|wxRIGHT|wxEXPAND, 5);

			wxBoxSizer *lnSzr=new wxBoxSizer(wxHORIZONTAL);
				lnSzr->AddSpacer(10);
				wxFlexGridSizer *flxSzr=new wxFlexGridSizer(2, 5, 5);

					m_chkInfos[CHK_APP_NAME] = new wxCheckBox(pnl, wxID_ANY, _T("Application's name"));
						m_chkInfos[CHK_APP_NAME]->Disable();
					flxSzr->Add(m_chkInfos[CHK_APP_NAME]);

					m_chkInfos[CHK_APP_VERSION] = new wxCheckBox(pnl, wxID_ANY, _T("Application's version"));
					flxSzr->Add(m_chkInfos[CHK_APP_VERSION]);

					m_chkInfos[CHK_APP_DESCRIPTION] = new wxCheckBox(pnl, wxID_ANY, _T("Description"));
					flxSzr->Add(m_chkInfos[CHK_APP_DESCRIPTION]);

					m_chkInfos[CHK_COPYRIGHT] = new wxCheckBox(pnl, wxID_ANY, _T("Copyright"));
					flxSzr->Add(m_chkInfos[CHK_COPYRIGHT]);

					m_chkInfos[CHK_ICON] = new wxCheckBox(pnl, wxID_ANY, _T("Icon"));
					flxSzr->Add(m_chkInfos[CHK_ICON]);

					m_chkInfos[CHK_LICENCE] = new wxCheckBox(pnl, wxID_ANY, _T("Licence text"));
					flxSzr->Add(m_chkInfos[CHK_LICENCE]);

					m_chkInfos[CHK_DEVELOPPER] = new wxCheckBox(pnl, wxID_ANY, _T("Main developper"));
					flxSzr->Add(m_chkInfos[CHK_DEVELOPPER]);

					m_chkInfos[CHK_DEVELOPPERS] = new wxCheckBox(pnl, wxID_ANY, _T("Multiple developpers"));
					flxSzr->Add(m_chkInfos[CHK_DEVELOPPERS]);

					m_chkInfos[CHK_WEBSITE] = new wxCheckBox(pnl, wxID_ANY, _T("Website url"));
					flxSzr->Add(m_chkInfos[CHK_WEBSITE]);

					m_chkInfos[CHK_TRANSLATORS] = new wxCheckBox(pnl, wxID_ANY, _T("Translators"));
					flxSzr->Add(m_chkInfos[CHK_TRANSLATORS]);

				lnSzr->Add(flxSzr, 0, wxALL, 0);
			stbSzr->Add(lnSzr, 0, wxALL, 5);
		szrMain->Add(stbSzr, 1, wxALL|wxEXPAND, 5);
	pnl->SetSizer(szrMain);

	// At startup, select the "All available infos" option
	m_optInfos[OPT_MAXIMAL]->SetValue(true);
	// Check the wxCheckBoxes depending on the global array (see on top of this file)
	for (size_t i=0; i<WXSIZEOF(g_iCheckedBoxes); ++i)
	{
		m_chkInfos[g_iCheckedBoxes[i]]->SetValue(true);
	}

	// Create the menu bar
	wxMenuBar *menuBar = new wxMenuBar();
		// File menu
		wxMenu *fileMenu = new wxMenu();
		fileMenu->Append(wxID_EXIT, wxGetStockLabel(wxID_EXIT, wxSTOCK_WITH_MNEMONIC|wxSTOCK_WITH_ACCELERATOR), wxGetStockHelpString(wxID_EXIT));
		menuBar->Append(fileMenu, wxGetStockLabel(wxID_FILE));
		// Help menu
		wxMenu* helpMenu = new wxMenu();
		helpMenu->Append(wxID_ABOUT, wxGetStockLabel(wxID_ABOUT, wxSTOCK_WITH_MNEMONIC|wxSTOCK_WITH_ACCELERATOR), wxGetStockHelpString(wxID_ABOUT));
		menuBar->Append(helpMenu, wxGetStockLabel(wxID_HELP));
	SetMenuBar(menuBar);
#ifdef __WXMAC__
	// On OSX, Exit and About menu entries are handled by the osx-specific menu
	// So the File and Help menus became empty : we can delete them
	menuBar->Remove(1); // Remove and the Help menu
	delete helpMenu;    // and delete it
	menuBar->Remove(0); // Remove the File menu
	delete fileMenu;    // and delete it
#endif // __WXMAC__

	szrMain->SetSizeHints(this);
}

void MainFrame::ConnectControls()
{
	Bind(wxEVT_MENU, &MainFrame::OnMenuExitClicked, this, wxID_EXIT);
	Bind(wxEVT_MENU, &MainFrame::OnMenuAboutClicked, this, wxID_ABOUT);
	Bind(wxEVT_RADIOBUTTON, &MainFrame::OnOptInfosClicked, this);
}

void MainFrame::UpdateInterface()
{
	// Checkboxes are only shown if the "custom" option is selected
	bool bShow = m_optInfos[OPT_CUSTOM]->GetValue();
	for (int i=0; i<CHK_COUNT; ++i)
		m_chkInfos[i]->Show(bShow);
}

void MainFrame::OnMenuExitClicked(wxCommandEvent &event)
{
	Destroy();
}

void MainFrame::OnMenuAboutClicked(wxCommandEvent &event)
{
	// Get the index of the selected option
	int iOpt=OPT_MAXIMAL;
	for (int i=0; i<OPT_COUNT; ++i)
	{
		if (m_optInfos[i]->GetValue())
			iOpt = i;
	}
	wxAboutDialogInfo info;
	wxIcon ico;

	for (int i=0; i<CHK_COUNT; ++i)
	{
		if ((i==CHK_APP_NAME)||(iOpt==OPT_MAXIMAL)||((iOpt==OPT_CUSTOM)&&(m_chkInfos[i]->GetValue())))
		{
			switch(i)
			{
				case CHK_APP_NAME:
					info.SetName(_T(PRODUCTNAME));
					break;
				case CHK_APP_VERSION:
					info.SetVersion(wxGetApp().GetVersionNumString(true));
					break;
				case CHK_APP_DESCRIPTION:
					info.SetDescription(wDESCRIPTION);
					break;
				case CHK_COPYRIGHT:
					info.SetCopyright(COPYRIGHT);
					break;
				case CHK_ICON:
					ico.CopyFromBitmap(wxGet_wxWidgets_png_Bitmap());
					info.SetIcon(ico);
					break;
				case CHK_LICENCE:
					info.SetLicence(g_szLicence);
					break;
				case CHK_DEVELOPPER:
					info.AddDeveloper(_T("X. Perrissoud AKA Xaviou"));
					break;
				case CHK_DEVELOPPERS:
					info.AddDeveloper(_T("An unknown developper"));
					info.AddDeveloper(_T("Another unknown developper"));
					break;
				case CHK_WEBSITE:
					info.SetWebSite(_T("https://wxstuff.xaviou.fr"), _T("X@v's wxStuff website"));
					break;
				case CHK_TRANSLATORS:
					info.SetTranslators(wxArrayString(WXSIZEOF(g_szTranslators), g_szTranslators));
					break;
			}
		}
	}

	wxAboutBox(info);
}
void MainFrame::OnOptInfosClicked(wxCommandEvent& event)
{
	UpdateInterface();
}

File: AboutBoxWx/Python/MainFrame.py
import wx
import wx.adv

class Options:
    MINIMAL = 0
    MAXIMAL = 1
    CUSTOM = 2

    COUNT = 3

class Details:
    APP_NAME = 0
    APP_VERSION = 1
    APP_DESCRIPTION = 2
    COPYRIGHT = 3
    ICON = 4
    LICENCE = 5
    DEVELOPPER = 6
    DEVELOPPERS = 7
    WEBSITE = 8
    TRANSLATORS = 9

    COUNT = 10

iCheckedBoxes = [
    Details.APP_NAME,
    Details.APP_VERSION,
    Details.APP_DESCRIPTION,
    Details.COPYRIGHT,
    Details.ICON
]

sLicence = '''This application is provided "as is" without any licence

Everyone is free to copy, modify, sell, and do anything with it.

This application and its source code is provided as a small
sample to show the use of wxWidgets standard about box.'''

aTranslators = ['Xaviou (en = app default)', 'Nobody Else (no translations support)']

class MainFrame(wx.Frame):

    def __init__(self):
        wx.Frame.__init__(self, None, title=wx.GetApp().GetAppName())
        self.SetIcon(wx.Icon('../../common/wxwin.ico'))

        self._optInfos = [0] * Options.COUNT
        self._chkInfos = [0] * Details.COUNT

        self._createControls()

        self._connectControls()

        self._updateInterface()

    def _createControls(self):
        # A Statusbar in the bottom of the window
        self.CreateStatusBar(1)
        sMsg = 'wxPython ' + wx.version() + ' - ' + wx.GetLibraryVersionInfo().VersionString
        self.SetStatusText(sMsg)

        # Add a panel to the frame (needed under Windows to have a nice background)
        pnl = wx.Panel(self, wx.ID_ANY)

        szrMain = wx.BoxSizer(wx.VERTICAL)
        stbSzr = wx.StaticBoxSizer(wx.VERTICAL, pnl, 'Informations to show in the "About" box:')

        self._optInfos[Options.MINIMAL] = wx.RadioButton(pnl, wx.ID_ANY, 'Application\'s name only (minimal "About" box)', style=wx.RB_GROUP)
        stbSzr.Add(self._optInfos[Options.MINIMAL], 0, wx.ALL, 5)

        self._optInfos[Options.MAXIMAL] = wx.RadioButton(pnl, wx.ID_ANY, 'All available informations')
        stbSzr.Add(self._optInfos[Options.MAXIMAL], 0, wx.LEFT|wx.RIGHT|wx.BOTTOM, 5)

        self._optInfos[Options.CUSTOM] = wx.RadioButton(pnl, wx.ID_ANY, 'Let me select the informations I want to see')
        stbSzr.Add(self._optInfos[Options.CUSTOM], 0, wx.LEFT|wx.RIGHT, 5)

        lnSzr = wx.BoxSizer(wx.HORIZONTAL)
        lnSzr.AddSpacer(10)

        flxSzr = wx.FlexGridSizer(2, 5, 5)

        self._chkInfos[Details.APP_NAME] = wx.CheckBox(pnl, wx.ID_ANY, 'Application\'s name')
        self._chkInfos[Details.APP_NAME].Disable()
        flxSzr.Add(self._chkInfos[Details.APP_NAME])

        self._chkInfos[Details.APP_VERSION] = wx.CheckBox(pnl, wx.ID_ANY, 'Application\'s version')
        flxSzr.Add(self._chkInfos[Details.APP_VERSION])

        self._chkInfos[Details.APP_DESCRIPTION] = wx.CheckBox(pnl, wx.ID_ANY, 'Description')
        flxSzr.Add(self._chkInfos[Details.APP_DESCRIPTION])

        self._chkInfos[Details.COPYRIGHT] = wx.CheckBox(pnl, wx.ID_ANY, 'Copyright')
        flxSzr.Add(self._chkInfos[Details.COPYRIGHT])

        self._chkInfos[Details.ICON] = wx.CheckBox(pnl, wx.ID_ANY, 'Icon')
        flxSzr.Add(self._chkInfos[Details.ICON])

        self._chkInfos[Details.LICENCE] = wx.CheckBox(pnl, wx.ID_ANY, 'Licence text')
        flxSzr.Add(self._chkInfos[Details.LICENCE])

        self._chkInfos[Details.DEVELOPPER] = wx.CheckBox(pnl, wx.ID_ANY, 'Main developper')
        flxSzr.Add(self._chkInfos[Details.DEVELOPPER])

        self._chkInfos[Details.DEVELOPPERS] = wx.CheckBox(pnl, wx.ID_ANY, 'Multiple developpers')
        flxSzr.Add(self._chkInfos[Details.DEVELOPPERS])

        self._chkInfos[Details.WEBSITE] = wx.CheckBox(pnl, wx.ID_ANY, 'Website url')
        flxSzr.Add(self._chkInfos[Details.WEBSITE])

        self._chkInfos[Details.TRANSLATORS] = wx.CheckBox(pnl, wx.ID_ANY, 'Translators')
        flxSzr.Add(self._chkInfos[Details.TRANSLATORS])

        lnSzr.Add(flxSzr, 0, wx.ALL, 0)

        stbSzr.Add(lnSzr, 0, wx.ALL, 5)

        szrMain.Add(stbSzr, 1, wx.ALL|wx.EXPAND, 5)

        pnl.SetSizer(szrMain)

        # At startup, select the "All available infos" option
        self._optInfos[Options.MAXIMAL].SetValue(True)
        # Check the wxCheckBoxes depending on the defined array (see on top of this file)
        for i in iCheckedBoxes:
            self._chkInfos[i].SetValue(True)

        # Create the menu bar
        menuBar = wx.MenuBar()
        # File menu
        menuFile = wx.Menu()
        menuFile.Append(wx.ID_EXIT)
        menuBar.Append(menuFile, wx.GetStockLabel(wx.ID_FILE))
        # Help menu
        menuHelp = wx.Menu()
        menuHelp.Append(wx.ID_ABOUT)
        menuBar.Append(menuHelp, wx.GetStockLabel(wx.ID_HELP))
        # Assigne the menubar
        self.SetMenuBar(menuBar)

        # On OS X, delete the created menus
        # as the About and Exit items are handled by the OS specific menu
        if wx.GetOsDescription()[:8] == 'Mac OS X':
            menuBar.Remove(1)
            del menuHelp
            menuBar.Remove(0)
            del menuFile

        # Adjust window size
        szrMain.SetSizeHints(self)

    def _connectControls(self):
        self.Bind(wx.EVT_MENU, self._onMenuExitClicked, id=wx.ID_EXIT)
        self.Bind(wx.EVT_MENU, self._onMenuAboutClicked, id=wx.ID_ABOUT)
        self.Bind(wx.EVT_RADIOBUTTON, self._onOptInfosClicked)

    def _updateInterface(self):
        # Checkboxes are only shown if the "custom" option is selected
        bShow = self._optInfos[Options.CUSTOM].GetValue()
        for chk in self._chkInfos:
            chk.Show(bShow)

    def _onMenuExitClicked(self, evt):
        self.Close()

    def _onMenuAboutClicked(self, evt):
        # Get the index of the selected option
        iOpt = Options.MAXIMAL
        for i in range(Options.COUNT):
            if self._optInfos[i].GetValue():
                iOpt = i
                break
        info = wx.adv.AboutDialogInfo()

        for i in range(Details.COUNT):
            if (i == Details.APP_NAME or iOpt == Options.MAXIMAL or
                (iOpt == Options.CUSTOM and self._chkInfos[i].GetValue())):

                if i == Details.APP_NAME:
                    info.SetName('AboutBoxWx')
                elif i == Details.APP_VERSION:
                    info.SetVersion('1.0.0')
                elif i == Details.APP_DESCRIPTION:
                    info.SetDescription('wxWidgets provided AboutBox')
                elif i == Details.COPYRIGHT:
                    info.SetCopyright('Copyright (c) X.P. 2018')
                elif i == Details.ICON:
                    info.SetIcon(wx.Icon('../../common/wxWidgets.png', wx.BITMAP_TYPE_PNG))
                elif i == Details.LICENCE:
                    info.SetLicence(sLicence)
                elif i == Details.DEVELOPPER:
                    info.AddDeveloper('X. Perrissoud AKA Xaviou')
                elif i == Details.DEVELOPPERS:
                    info.AddDeveloper('An unknown developper')
                    info.AddDeveloper('Another unknown developper')
                elif i == Details.WEBSITE:
                    info.SetWebSite('https://wxstuff.xaviou.fr', 'X@v\'s wxStuff website')
                elif i == Details.TRANSLATORS:
                    info.SetTranslators(aTranslators)

        wx.adv.AboutBox(info)

    def _onOptInfosClicked(self, evt):
        self._updateInterface()
